File: Pseudo-boolean.cpp
std::map<std::pair<int,int>,int> VMCProblem::PBConstraintLEQ(std::vector<int> vars, std::vector<int> weights, int k)
{
    std::map<std::pair<int,int>,int> s_ij =  std::map<std::pair<int,int>,int>(); //aux variables
    std::vector<int> literals = std::vector<int>();

    //Create all auxiliary vars
    for(int i=0; i < vars.size(); i++)
        for(int j=1; j <= k; j++ )
            s_ij[std::pair<int, int>(i, j)] = _varDispatcher->genAux();
    _SATSolver->addVars((int) vars.size() * k);

    //Remove literals whose coefficients are greater than k, and assign them value 0
    std::vector<int> auxVars;
    std::vector<int> auxWeights;
    for(int i=0; i < vars.size(); i++)
    {
        if(weights[i] > k)
        {
            literals.clear();
            literals.push_back(-vars[i]);
            _SATSolver->addClause(literals);
        }
        else
        {
            auxVars.push_back(vars[i]);
            auxWeights.push_back(weights[i]);
        }
    }
    vars = auxVars;
    weights = auxWeights;




    for(int j=1; j <= weights[0]; j++)
    {
        literals.clear();
        literals.reserve(2);

        literals.push_back(-vars[0]);
        literals.push_back(s_ij[std::pair<int,int>(0,j)]);

        _SATSolver->addClause(literals);                                // (-x1 v S1j)
    }


    for(int j=weights[0]+1; j <= k; j++)
    {
        literals.clear();
        literals.reserve(1);

        literals.push_back(- s_ij[std::pair<int,int>(0,j)]);

        _SATSolver->addClause(literals);                                // (-S1j)
    }


    for(int i=1; i < vars.size(); i++)
    {
        for(int j=1; j <= weights[i]; j++)
        {
            literals.clear();
            literals.reserve(2);

            literals.push_back(-vars[i]);
            literals.push_back(s_ij[std::pair<int,int>(i,j)]);

            _SATSolver->addClause(literals);                             // (-xi v Sij)

        }


        for(int j=1; j <= k; j++)
        {
            literals.clear();
            literals.reserve(2);

            literals.push_back(- s_ij[std::pair<int,int>(i-1,j)]);
            literals.push_back(s_ij[std::pair<int,int>(i,j)]);

            _SATSolver->addClause(literals);                             // (-Si-1,j v Sij)
        }

        for(int j=1; j <= k - weights[i]; j++)
        {
            literals.clear();
            literals.reserve(3);

            literals.push_back(- vars[i]);
            literals.push_back(- s_ij[std::pair<int,int>(i-1,j)]);
            literals.push_back(s_ij[std::pair<int,int>(i, j + weights[i] )]);

            _SATSolver->addClause(literals);                             // (-xi v -Si-1,j v Si,j+wi)
        }

        {
            literals.clear();
            literals.reserve(2);

            literals.push_back(- vars[i]);
            literals.push_back(- s_ij[std::pair<int,int>(i-1, k+1-weights[i] )]);

            _SATSolver->addClause(literals);                             // (-xi v -Si-1,k+1-wi)
        }
    }

    return s_ij;
}


/**
 *
 * Encodes a Pseudo Boolean constraint of type w1.x1 + w2.x2 + ... >= k
 */
std::map<std::pair<int,int>,int> VMCProblem::PBConstraintGEQ(std::vector<int> vars, std::vector<int> weights, int k)
{

    int newK = 0;

    for(int i=0; i < weights.size(); i++)
    {
        vars[i] = -vars[i];
        newK += weights[i];
    }
    newK -= k;

    return this->PBConstraintLEQ(vars, weights, newK);
}

File: pseudo_boolean.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Feb 11 14:54:41 2018

@author: RitaRamos
"""

import var_despatcher

def PBConstraintLEQ(my_vars,weights, k):
    s_ij={} 
    literals =[]
    vars_size=len(my_vars)
    
    for i in range(vars_size):
        for j in range (1, k + 1):
            s_ij[(i,j)]= var_despatcher.gen_aux() #GONCALLOOOO
        
    auxVars=[]
    auxWeights=[]
    
    for i in range(my_vars):
        if weights[i]>k:
            _SATSolver.add_clause([-my_vars[i]])
        else:
            auxVars+=my_vars[i]
            auxWeights+=weights[i]
    my_vars=auxVars
    weights=auxWeights
    
    
    for j in range(1, weights[0] +1):
        _SATSolver.add_clause([-my_vars[0], s_ij[(0,j)] ])
    
    for j in range(weights[0]+1, k +1):
        _SATSolver.add_clause([-s_ij[(0,j)] ])
    
    for i in range(1, len(my_vars)):
        for j in range(1, weights[i]):
            _SATSolver.add_clause( [-my_vars[i], s_ij[(i,j)] ])
            
    for j in range(1, k+1):
        _SATSolver.add_clause( [ s_ij[(i-1,j)]    , s_ij[(i,j)] ])
        
    for j in range(1, k- weights[i]+ 1):
        _SATSolver.add_clause( [ -my_vars[i], s_ij[(i-1,j)]    , s_ij[(i,j +weights[i])] ])
    
    _SATSolver.add_clause( [-my_vars[i], s_ij[(i-1, k+1-weights[i] )] ])
    
    return s_ij


def PBConstraintGEQ(my_vars, weights,  k):
    newK = 0
    
    for i in range(len(weights)):
        my_vars[i] = -my_vars[i]
        newK += weights[i]

    newK -= k

    return PBConstraintLEQ(my_vars, weights, newK)       
            
            
